### ISSUE TITLE
dotnet publish does not respect target framework from publish profile

### ISSUE THREAD

**Issue by holc-tkomp** [dotnet/sdk#11474]
## Description
Running `dotnet publish -p:PublishProfile=Anything.pubxml` ignores the fact that the `TargetFramework` is specified within given `.pubxml` file and informs you (displays error) that you must specify the framework manually.

## Current behavior
Error is displayed and publish process is stopped
> The 'Publish' target is not supported without specifying a target framework. The current project targets multiple frameworks, please specify the framework for the published application.

## Expected behavior
`TargetFramework` should be retrieved from the `.pubxml` file.

## Steps to reproduce
1. `dotnet new console`.
2. In `.csproj` set `TargetFrameworks` to `netcoreapp2.1;netcoreapp3.1`.
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFrameworks>netcoreapp2.1;netcoreapp3.1</TargetFrameworks>
    <RootNamespace>dotnet_publish_issue</RootNamespace>
  </PropertyGroup>
</Project>
```
3. Open up with VS, create any folder publish profile for `netcoreapp2.1`. Created profile should look like this
```xml
<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <PropertyGroup>
    <PublishProtocol>FileSystem</PublishProtocol>
    <Configuration>Release</Configuration>
    <Platform>Any CPU</Platform>
    <TargetFramework>netcoreapp2.1</TargetFramework>
    <PublishDir>bin\Release\netcoreapp2.1\publish\</PublishDir>
  </PropertyGroup>
</Project>
```
4. Executing `dotnet publish -p:PublishProfile=Properties\PublishProfiles\FolderProfile.pubxml` results in error:
```wrap
C:\Program Files\dotnet\sdk\3.1.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.CrossTargeting.targets(27,5): error : The 'Publish' target is not supported without specifying a target framework. The current project targets multiple frameworks, please specify the framework for the published application. [C:\dotnet publish issue\dotnet publish issue.csproj]
```

Executing `dotnet publish -p:PublishProfile=Properties\PublishProfiles\FolderProfile.pubxml -f:netcoreapp2.1` works fine, but why does `.pubxml`'s `<TargetFramework>` node is ignored?

**Comment by dsplaisted** [comment]
This is happening because the publish profile isn't even imported if the `TargetFramework` is not specified.  When `TargetFramework` is not specified, we import `Microsoft.NET.Sdk.BeforeCommonCrossTargeting.targets` instead of `Microsoft.NET.Sdk.BeforeCommon.targets`, which is what imports the publish profile.

@vijayrkn, do you think it's feasible to fix this?

**Comment by vijayrkn** [comment]
Looks like the change we are talking about here is to add this one line https://github.com/dotnet/sdk/blob/28b3115b58670323d81372b31f0348eadf4bfc50/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommon.targets#L40-L44

here right?https://github.com/dotnet/sdk/blob/ab2176b3447e3e7820b5a340408110b6d498f907/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.BeforeCommonCrossTargeting.targets#L16

Doesn't look like a costly change. I can test it out.

**Comment by vijayrkn** [comment]
Ok the change is more involved than just importing the publish target. Lot of dependent targets are not imported like these. 
`The target "ResolveLockFileCopyLocalFiles" does not exist in the project.`

**Comment by DNF-SaS** [comment]
Seems like this has been fixed for at least SDK 5.0.200-preview.20614.14 ?

**Comment by danielgindi** [comment]
`RuntimeIdentifier` is ignored too. Defaults to current platform. Only explicitly setting `-r` changes the runtime. This is very misleading when using publish from within VS, explicitly setting `RuntimeIdentifier` to linux and getting a windows exe.

**Comment by dekenless** [comment]
I'm still seeing this in .NET 6

**Comment by Elangovan4ever** [comment]
This issue still exists in .NET 7

**Comment by baronfel** [comment]
Yep, we need to do an evaluation and then trigger the Publish with the properties from the selected PublishProfile, which is exactly what VS does. I call this a 'two-phase' publish, and I have a draft PR for this [here](https://github.com/dotnet/sdk/pull/29817).  I need to get feedback from the VS Publish folks on the exact set of properties that are lifted from the PublishProfile, and write tests, but I am halfway working on proposing this for 7.0.300.

The reason this two-phase approach is needed is because of MSBuild props/targets ordering - the PublishProfile evaluation is done long after certain 'core' properties like the RID, TFM, etc are set - and in fact many of these properties can be set by the CLI itself. When this occurs they are 'global properties' for the build, and as such can't be overwritten (which is what a naieve Import of the Publish Profile would do).

**Comment by mic-code** [comment]
Encountered this just now, I was using both visual studio and CI to publish, needed to define stuff twice in a publish profile and the CI script

**Comment by baronfel** [comment]
I had an attempt at this in https://github.com/dotnet/sdk/pull/29817, but it needs to be reworked a bit.
